### PR TITLE
[JENKINS-26130] Way to print progress from pickles

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/5, https://github.com/jenkinsci/workflow-support-plugin/pull/4, https://github.com/jenkinsci/workflow-job-plugin/pull/5, and https://github.com/jenkinsci/workflow-cps-plugin/pull/20.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java [hunk 1]
```diff
@@ -60,21 +67,29 @@ private ExecutorPickle(Executor executor) {
...
                     if (item == null) {
                         // TODO should also report when !ScheduleResult.created, since that is arguably an error
                         throw new IllegalStateException("queue refused " + task);
                     }
-                    future = item.getFuture().getStartCondition();
+                    itemID = item.getId();
+                } else {
+                    item = Queue.getInstance().getItem(itemID);
+                    if (item == null) {
+                        throw new IllegalStateException("queue lost item #" + itemID);
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java]: I'm not sure if it's safe if the QueueTask gets completed quickly

File: src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java [hunk 2]
```diff
@@ -88,6 +103,33 @@ protected Executor tryResolve() throws Exception {
...
+                }
+            }
+            @Override public boolean cancel(boolean mayInterruptIfRunning) {
+                Queue.Item item = Queue.getInstance().getItem(itemID);
+                if (item != null) {
+                    if (!Queue.getInstance().cancel(item)) {
+                        LOGGER.log(Level.WARNING, "failed to cancel {0}", item);
+                    }
+                }
+                return super.cancel(mayInterruptIfRunning);
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java]: Is it really required?

File: src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java]: `Queue.leftItems` keeps things for five minutes, much longer than the recurrence period for this pickle.

File: src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java]: Which part? The `Queue.cancel` call removes the _part of …_ task from the queue, which the last line of `unloadableExecutorPickle` asserts. The `super.cancel` stops the retrying in `TryRepeatedly` (though the next try would probably fail anyway, in `QueueTaskFuture.get`).

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/support/pickles/ExecutorPickle.java]: ok

**Comment by oleg-nenashev** [comment]
🐝

**Comment by abayer** [comment]
To paraphrase Eddie Izzard, "I like my pull requests like I like my coffee - covered in :bee:s"

**Pull request merged**